File: src/System.Commands/ICommandResolver.vb
''' <summary>
''' Command resolver interface for resolving commands.
''' This interface defines methods to resolve commands based on their names or types.
''' </summary>
''' <remarks>
''' Implement this interface to create custom command resolvers that can resolve commands in a specific context.
''' </remarks>
Public Interface ICommandResolver
    ''' <summary>
    ''' Clears the command resolver.
    ''' This method should clear any cached or stored commands in the resolver.
    ''' It can be used to reset the state of the resolver, for example, when reloading commands or changing contexts.
    ''' </summary>
    Sub Clear()

    ''' <summary>
    ''' Resolves a command by its type.
    ''' </summary>
    ''' <typeparam name="TCommand">The type of the command to resolve.</typeparam>
    ''' <returns>The command instance associated with the specified type, or a new instance if not found.</returns>
    Function Resolve(Of TCommand As {Class, ICommand, New})() As TCommand

    ''' <summary>
    ''' Registers a command factory for a specific command type.
    ''' This method allows you to register a factory function that creates instances of the specified command type.
    ''' The factory will be used to create new instances of the command when it is resolved.
    ''' </summary>
    ''' <typeparam name="TCommand">The type of the command to register.</typeparam>
    ''' <param name="factory">A method to create a new TCommand object.</param>
    ''' <returns>An object that can be used to un-register</returns>
    Function RegisterCommand(Of TCommand As {Class, ICommand, New})(factory As Func(Of TCommand)) As IDisposable

    ''' <summary>
    ''' Registers a command factory for a specific command type.
    ''' This method allows you to register a factory function that creates instances of the specified command type.
    ''' The factory will be used to create new instances of the command when it is resolved.
    ''' </summary>
    ''' <typeparam name="TCommand">The type of the command to register.</typeparam>
    ''' <param name="parentType"> The type of the parent command that the command is associated with.</param>
    ''' <param name="factory">A method to create a new TCommand object.</param>
    ''' <returns>An object that can be used to un-register</returns>
    Function RegisterCommand(Of TCommand As {Class, ICommand, New})(parentType As Type, factory As Func(Of TCommand)) As IDisposable

End Interface

File: src/Commands/ICommandResolver.cs
namespace Commands
{
	/// <summary>
	/// Resolves, registers and manages command factory registrations used to create <see cref="ICommand"/> instances.
	/// </summary>
	/// <remarks>
	/// Implementations provide a way to register factories for specific command types and to resolve new command instances
	/// using those factories. Registrations return an <see cref="IDisposable"/> which should remove the registration when disposed.
	/// </remarks>
	public interface ICommandResolver
	{
		/// <summary>
		/// Clears all registered command factories.
		/// </summary>
		void Clear();

		/// <summary>
		/// Resolves a new instance of the requested command type.
		/// </summary>
		/// <typeparam name="TCommand">The concrete command type to resolve. Must implement <see cref="ICommand"/>, be a reference type and have a public parameterless constructor.</typeparam>
		/// <returns>A new instance of <typeparamref name="TCommand"/> created by the registered factory or default factory.</returns>
		TCommand Resolve<TCommand>()
			where TCommand : class, ICommand, new();

		/// <summary>
		/// Registers a factory that creates instances of <typeparamref name="TCommand"/>.
		/// </summary>
		/// <typeparam name="TCommand">The command type the factory produces. Must implement <see cref="ICommand"/>, be a reference type and have a public parameterless constructor.</typeparam>
		/// <param name="factory">A factory delegate that returns new instances of <typeparamref name="TCommand"/>.</param>
		/// <returns>
		/// An <see cref="IDisposable"/> which, when disposed, removes the registration for the provided command type.
		/// </returns>
		IDisposable Register<TCommand>(Func<TCommand> factory)
			where TCommand : class, ICommand, new();

		/// <summary>
		/// Registers a factory that creates instances for the specified command <paramref name="commandType"/>.
		/// </summary>
		/// <param name="commandType">The runtime <see cref="Type"/> of command the factory produces. Must implement <see cref="ICommand"/>.</param>
		/// <param name="factory">A factory delegate that returns new instances of <see cref="ICommand"/>.</param>
		/// <returns>
		/// An <see cref="IDisposable"/> which, when disposed, removes the registration for the specified <paramref name="commandType"/>.
		/// </returns>
		IDisposable Register(Type commandType, Func<ICommand> factory);
	}
}
